File: csharp/annotations/annotation_replies/annotation_replies.cs
//  annotation_replies -- C# port of examples\Vb6\annotations\annotation_replies\annotation_replies.bas
//  A square annotation with a reply, and a reply to that reply.
using System;
using System.IO;
using LumasPdfSdk;

class AnnotationReplies
{
    const uint NO_COLOR = 0xFFFFFFF1;   // transparent

    static readonly TErrorProc _err = PDFError;
    static int PDFError(IntPtr Data, int ErrCode, string ErrMessage, int ErrType) { return 0; }

    static void Main()
    {
        IntPtr pdf = LumasPdf.pdfNewPDF();

        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _err);
        LumasPdf.pdfCreateNewPDFW(pdf, "");

        LumasPdf.pdfSetPageCoords(pdf, (int)TPageCoord.pcTopDown);

        LumasPdf.pdfAppend(pdf);
        int annot = LumasPdf.pdfSquareAnnotW(pdf, 50, 50, 200, 100, 3, NO_COLOR, 255, TPDFColorSpace.csDeviceRGB, "Jim", "Test", "Just test...");
        int reply = LumasPdf.pdfSetAnnotMigrationStateW(pdf, (uint)annot, TAnnotState.asCreateReply, "Harry");
        LumasPdf.pdfSetAnnotStringW(pdf, (uint)reply, (int)TAnnotString.asContent, "This is a reply!");

        reply = LumasPdf.pdfSetAnnotMigrationStateW(pdf, (uint)reply, TAnnotState.asCreateReply, "Jim");
        LumasPdf.pdfSetAnnotStringW(pdf, (uint)reply, (int)TAnnotString.asContent, "This is a reply to a reply!");
        LumasPdf.pdfEndPage(pdf);

        if (LumasPdf.pdfHaveOpenDoc(pdf))
        {
            string outFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "out.pdf");
            if (!LumasPdf.pdfOpenOutputFileW(pdf, outFile)) { LumasPdf.pdfDeletePDF(pdf); return; }
            if (LumasPdf.pdfCloseFile(pdf))
                Console.WriteLine("PDF file \"" + outFile + "\" successfully created!");
        }

        LumasPdf.pdfDeletePDF(pdf);
    }
}

File: c/annotations/annotation_replies/annotation_replies.c
/* annotation_replies -- C (x64) port of examples\Vb6\annotations\annotation_replies.bas */
#include <stdio.h>
#include <string.h>
#include "lumaspdf.h"

#define NO_COLOR 0xFFFFFFF1u   /* transparent */

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType)
{
    (void)Data; (void)ErrCode; (void)ErrType;
    if (ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

static void exedir(const char* a0, char* out, size_t n){char* s;strncpy(out,a0,n-1);out[n-1]=0;s=strrchr(out,'\\');if(!s)s=strrchr(out,'/');if(s)*s=0;else strcpy(out,".");}

int main(int argc, char** argv)
{
    SI32 annot, reply;
    PPDF pdf;
    char dir[1024], outFile[1200];

    exedir(argv[0], dir, sizeof(dir));
    (void)argc;

    pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, ErrProc);
    pdfCreateNewPDFA(pdf, "");

    pdfSetPageCoords(pdf, pcTopDown);

    pdfAppend(pdf);
    annot = pdfSquareAnnotA(pdf, 50.0, 50.0, 200.0, 100.0, 3.0, NO_COLOR, 255, csDeviceRGB, "Jim", "Test", "Just test...");
    reply = pdfSetAnnotMigrationStateA(pdf, annot, asCreateReply, "Harry");
    pdfSetAnnotStringA(pdf, reply, asContent, "This is a reply!");

    reply = pdfSetAnnotMigrationStateA(pdf, reply, asCreateReply, "Jim");
    pdfSetAnnotStringA(pdf, reply, asContent, "This is a reply to a reply!");
    pdfEndPage(pdf);

    if (pdfHaveOpenDoc(pdf) != 0) {
        _snprintf(outFile, sizeof(outFile), "%s\\out.pdf", dir);
        if (pdfOpenOutputFileA(pdf, outFile) == 0) {
            pdfDeletePDF(pdf);
            return 0;
        }
        if (pdfCloseFile(pdf) != 0)
            printf("PDF file \"%s\" successfully created!\n", outFile);
    }

    pdfDeletePDF(pdf);
    return 0;
}
